client/trino-client/src/main/java/io/trino/client/uri/TrinoUri.java

```diff
@@ -0,0 +1,1196 @@
+        this.externalAuthentication = EXTERNAL_AUTHENTICATION.getValue(properties);
+        this.externalAuthenticationTimeout = EXTERNAL_AUTHENTICATION_TIMEOUT.getValue(properties);
+        this.externalRedirectStrategies = EXTERNAL_AUTHENTICATION_REDIRECT_HANDLERS.getValue(properties);
+        this.externalAuthenticationTokenCache = EXTERNAL_AUTHENTICATION_TOKEN_CACHE.getValue(properties);
+        this.extraCredentials = EXTRA_CREDENTIALS.getValue(properties);
+        this.clientInfo = CLIENT_INFO.getValue(properties);
+        this.clientTags = CLIENT_TAGS.getValue(properties);
+        this.traceToken = TRACE_TOKEN.getValue(properties);
+        this.sessionProperties = SESSION_PROPERTIES.getValue(properties);
+        this.source = SOURCE.getValue(properties);
+    }
+
+    private void initCatalogAndSchema()
+            throws SQLException
+    {
+        String path = uri.getPath();
+        if (isNullOrEmpty(uri.getPath()) || path.equals("/")) {
+            return;
+        }
+
+        // remove first slash
+        if (!path.startsWith("/")) {
+            throw new SQLException("Path does not start with a slash: " + uri);
+        }
+        path = path.substring(1);
+
+        List<String> parts = Splitter.on("/").splitToList(path);
+        // remove last item due to a trailing slash
+        if (parts.get(parts.size() - 1).isEmpty()) {
+            parts = parts.subList(0, parts.size() - 1);
+        }
+
+        if (parts.size() > 2) {
+            throw new SQLException("Invalid path segments in URL: " + uri);
+        }
+
+        if (parts.get(0).isEmpty()) {
+            throw new SQLException("Catalog name is empty: " + uri);
+        }
+
+        if (catalog.isPresent()) {
+            throw new RestrictedPropertyException(PropertyName.CATALOG, "Catalog cannot be set in the URL");
+        }
+        catalog = Optional.ofNullable(parts.get(0));
+
+        if (parts.size() > 1) {
+            if (parts.get(1).isEmpty()) {
+                throw new SQLException("Schema name is empty: " + uri);
+            }
+
+            if (schema.isPresent()) {
+                throw new RestrictedPropertyException(PropertyName.SCHEMA, "Schema cannot be set in the URL");
+            }
+            schema = Optional.ofNullable(parts.get(1));
+        }
+    }
+
+    private Properties mergeConnectionProperties(URI uri, Properties driverProperties)
+            throws SQLException
+    {
+        Map<String, String> urlProperties = parseParameters(uri.getQuery());
+        Map<String, String> suppliedProperties = Maps.fromProperties(driverProperties);
+
+        for (String key : urlProperties.keySet()) {
+            if (suppliedProperties.containsKey(key)) {
+                throw new SQLException(format("Connection property '%s' is both in the URL and an argument", key));
+            }
+        }
+
+        Properties result = new Properties();
+        setProperties(result, suppliedProperties);
```

Comment:
Make sure there's a test to check that properties take precedence over URL parameters.